### PR TITLE
Build: Move monorepo to Yarn 2 

### PR DESCRIPTION
**Pull request by gaetanmaisse**
Applying https://github.com/storybookjs/storybook/pull/13559 step by step to not face gigantic conflicts or issues with the webpack 4/5 works.

## What I did

> Upgrading the Yarn binary to the latest stable release (2.4), which should make them faster.
> 
> I've configured the project to stay on the node_modules linker since it makes the upgrade fairly painless. Tests pass locally, I'll see how it goes on CI and tweak the command line references accordingly (things like --frozen-lockfile being renamed to --immutable, etc).
>
> Notable changes:

> The Yarn binary is checked within this repository. This means that all your contributors will now be ensured to use the exact same version of Yarn, regardless of what they have installed globally.
>
> The root package.json now lists all workspaces as dependencies (using the workspace: protocol). This is because in Yarn 2 the root workspace doesn't implicitly depend on all workspaces anymore, they have to be explicitly listed. In theory, the dependencies should be listed in each workspace that depends on them (so, for example, ui should depend on api) but I'm worried it would lead to a gigantic diff, so, for now, the repo keeps relying on the hoisting to work.

## How to test

- CI should be 🟢 
- To test things locally, you should be able to do the following: 
  - Checkout this branch
  - `yarn install`
  - `yarn bootstrap --core`
  - `yarn test`
  - Go in any example and run `yarn storybook`

### CODE REVIEW

File: examples/angular-cli/.gitignore
Comment (merceyz) [inline examples/angular-cli/.gitignore:44]:
```suggestion
/.yarn/*
!/.yarn/releases
!/.yarn/plugins
!/.yarn/sdks
!/.yarn/versions
/.pnp.*
```

https://yarnpkg.com/getting-started/qa/#which-files-should-be-gitignored

Comment (gaetanmaisse) [reply examples/angular-cli/.gitignore]: And it was the wrong file 🙃  I did pay attention to when I cherry-picked that from the other branch.

**Comment by gaetanmaisse** [comment]
I mainly applied what @arcanis did in #13559 and did a bit of magic to make the E2E tests work with Yarn 2.
Many of them are now using `yarn dlx` instead of `npx`, the remaining ones need a deeper investigation that will be done when we will rework the whole E2E tests process to leverage the use of Yarn 2.

File: examples/ember-cli/package.json
Comment (gaetanmaisse) [inline examples/ember-cli/package.json:8]: to rename to `storybook`

File: scripts/run-e2e.ts
Comment (gaetanmaisse) [inline scripts/run-e2e.ts:257]: Explain why we need yarn2

Comment (gaetanmaisse) [reply scripts/run-e2e.ts:257]: Copy paste from the monorepo

**Comment by ndelangen** [comment]
Amazing work! 👏

**Comment by phated** [comment]
@gaetanmaisse One thing I've found is that I can't do `yarn bootstrap --reset` when checking out your PR for the first time. Will this cause any issues for maintainers' workflows once we merge this? Maybe I just need to do a `yarn install` before resetting, but that feels a little counter-intuitive

**Comment by gaetanmaisse** [comment]
@phated Hmm I see, I need to confess I have never used `yarn bootstrap --reset` before seeing your comment 🙈  
Sadly I'm not sure we can do much about that and I guess trying to `yarn install` when you face a weird error is now a reflex for many web developer (I'm not telling it's a good thing tho 😁 )

**Comment by arcanis** [comment]
I'm not sure what `bootstrap --reset` does (I don't see it in [the doc?](https://github.com/lerna/lerna/tree/main/commands/bootstrap#readme)) but, assuming it's to make sure the installed deps are the right ones for the current branch, a `yarn install` is all that should be needed to sync the state. If it isn't, it's worth opening an issue on our side.

**Comment by gaetanmaisse** [comment]
@arcanis it's part of the Storybook monorepo [internal cli](https://github.com/storybookjs/storybook/blob/b75fd745dfe825b49218dca3db7025e3e5cb0819/scripts/bootstrap.js), I don't think it worth to spend time on that, a `yarn install` and people will be good to go.

**Comment by phated** [comment]
@arcanis `bootstrap` is just an internal script for maintainers and `--reset` cleans out all artifacts (typescript generated files and all the `node_modules` throughout the monorepo). I run this each morning to make sure the project is in a clean state and I don't have weird dependency or typescript issues. I don't think we will have any issues once switching over, but I wanted  to flag that this might cause confusion on the first time when it is merged.

**Comment by shilman** [comment]
I use `--reset` every time I publish. I don't mind running a separate CLI to get the current `--reset` behavior (i.e. it doesn't need to be part of bootstrap), but that functionality should be part of the monorepo in one way or another. It's too easy for node_modules to get corrupted when switching between branches.

**Comment by phated** [comment]
@shilman sorry for being confusing. The `--reset` flag still works, just that you need to have yarn2's configuration "set up" for a first time before it works. I guess that could be an issue when switching between old pre-yarn2 branches and new post-yarn2 branches.

**Review by phated — APPROVED** [review]
I **love** how straight-forward this PR is because all the smaller changes have already been upstreamed! Great work @gaetanmaisse!!

**Comment by tooppaaa** [comment]
Except `yarn start` from the root that fail, everything is fine on windows!

**Comment by merceyz** [comment]
> Except `yarn start` from the root that fail, everything is fine on windows!

There is a bug open for that on our side at https://github.com/yarnpkg/berry/issues/2346.

Workaround:
```sh
yarn ./examples/official-storybook storybook --no-manager-cache
# or
yarn workspace official-storybook storybook --no-manager-cache
```

**Comment by gaetanmaisse** [comment]
Yes I picked `yarn workspace official-storybook storybook --no-manager-cache` I think it's even better than the `--cwd`

File: examples/html-kitchen-sink/.yarn/build-state.yml
Comment (merceyz) [inline examples/html-kitchen-sink/.yarn/build-state.yml:4]: Should remove and gitignore these

**Review by merceyz — COMMENTED** [review]
Keep in mind that the `node-modules` linker can't really install portals, once https://github.com/yarnpkg/berry/pull/2481 is merged it will try its best but it can only do so much with the filesystem.

File: .circleci/config.yml
Comment (merceyz) [inline .circleci/config.yml:98]: Yarn wont clear the global cache unless you tell it to so this will grow forever, I suggest setting
```
YARN_ENABLE_GLOBAL_CACHE=false
YARN_ENABLE_MIRROR=false
```
as env variables and caching `.yarn/cache` instead, which we do clean.

**Comment by gaetanmaisse** [comment]
@merceyz yep... I'm playing with things to try to move the examples out of the core/lib/addons workspaces but I think I will need to discuss with you to find out what we can really do and the things you have in mind that would help decrease our pains related to hoisting 😇

**Comment by gaetanmaisse** [comment]
Note: Need to rework the structure of the E2E (+ everything is running on Node 12 now), a better split could be:
 - `e2e-tests-core` for "officially" supported frameworks : React + Vue + Angular + Web Components (could be the future `@storybook/lit`, in PnP mode by default)
 - `e2e-tests-extended` for other frameworks: Preact, Svelte, Riot, Rax etc 
 - `e2e-tests-pnp` for some frameworks/setups compatible with PnP: React + WC --> End goal would be to use PnP as default mode for all compatible setups

**Review by shilman — APPROVED** [review]
Hallelujah!!! 🙌

**Pull request merged**
